### PR TITLE
Optimize PathGradientBrush, cleanup InternalPath, share ThreadLocal code

### PR DESCRIPTION
**Pull request by antonfirsov**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp.Drawing/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Follow up on https://github.com/SixLabors/ImageSharp.Drawing/pull/128#issuecomment-827738628 and https://github.com/SixLabors/ImageSharp.Drawing/pull/128#discussion_r622002284:
- Simplify and speed up `PathGradientBrush` with basic intersection code. (~5 speedup on a basic benchmark)
- Remove `IPath.Contains`. Although did not mention in #103 it also solves computational geometry concern which is unscoped, and current implementation relied on `InternalPath.FindIntersections`
- Remove `InternalPath.FindIntersections` and all related code
- Shared utility for thread local blender buffers

#### Benchmarks

Before:
```
|                       Method |     Mean |     Error |   StdDev |   Median |
|----------------------------- |---------:|----------:|---------:|---------:|
| FillGradientBrush_ImageSharp | 9.726 ms | 0.3696 ms | 1.030 ms | 9.368 ms |
```

After:
```
|                       Method |     Mean |     Error |    StdDev |
|----------------------------- |---------:|----------:|----------:|
| FillGradientBrush_ImageSharp | 2.144 ms | 0.0769 ms | 0.2206 ms |
```

### CODE REVIEW

**Comment by antonfirsov** [comment]
ImageMagick hates my new reference image:
https://github.com/SixLabors/ImageSharp.Drawing/blob/0da47ef2ee5b3772c48324ecdaf01178751f8cd8/tests/Images/ReferenceOutput/Drawing/GradientBrushes/FillPathGradientBrushTests/FillComplex.png

```
Error Message:
   ImageMagick.MagickMissingDelegateErrorException : NoDecodeDelegateForThisImageFormat `' @ error/blob.c/CustomStreamToImage/804
```

@dlemstra any idea? Is the version up-to-date in this repo?

https://github.com/SixLabors/ImageSharp.Drawing/blob/d8960b8e2e55875d2c4f990f5fc439b846af1230/tests/Directory.Build.targets#L22

**Comment by JimBobSquarePants** [comment]
@antonfirsov It'll be LFS. Looks like new images are being automatically added due to the new editorconfig. 

You need to enable LFS in the workflow and also enable caching. 

Copy the yml from ImageSharp, update the infrastructure submodule, and change the `XUNIT_PATH` path variable and all should be ok. 

https://github.com/SixLabors/ImageSharp/blob/a6c17854684edaa998080233273d19d987174bb5/.github/workflows/build-and-test.yml

I can do this if you want for ease?

File: src/ImageSharp.Drawing/Shapes/InternalPath.cs [hunk 1]
```diff
@@ -117,345 +117,6 @@ private InternalPath(PointData[] points, bool isClosedPath)
         /// </summary>
         public int PointCount => this.points.Length;
 
-        /// <summary>
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Drawing/Shapes/InternalPath.cs:120]: This fills my heart with joy!

File: src/ImageSharp.Drawing/Utilities/Intersect.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace SixLabors.ImageSharp.Drawing.Utilities
+{
+    internal static class Intersect
+    {
+        private const float Eps = 1e-3f;
+        private const float MinusEps = -Eps;
+        private const float OnePlusEps = 1 + Eps;
+
+        public static bool LineSegmentToLineSegmentIgnoreCollinear(Vector2 a0, Vector2 a1, Vector2 b0, Vector2 b1, ref Vector2 intersectionPoint)
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Drawing/Utilities/Intersect.cs:14]: It blows my mind how simple this is. Not something I'd have ever figured out though!

File: src/ImageSharp.Drawing/Utilities/ThreadLocalBlenderBuffers.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace SixLabors.ImageSharp.Drawing.Utilities
+{
+    internal class ThreadLocalBlenderBuffers<TPixel> : IDisposable
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        private ThreadLocal<BufferOwner> data;
+
+        // amountBufferOnly:true is for SolidBrush, which doesn't need the overlay buffer (it will be dummy)
+        public ThreadLocalBlenderBuffers(MemoryAllocator allocator, int scanlineWidth, bool amountBufferOnly = false)
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Drawing/Utilities/ThreadLocalBlenderBuffers.cs:18]: Was curious how you could handle the difference.

File: src/ImageSharp.Drawing/Utilities/ThreadLocalBlenderBuffers.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            foreach (BufferOwner d in this.data.Values)
+            {
+                d.Dispose();
+            }
+
+            this.data.Dispose();
+            this.data = null;
+        }
+
+        private sealed class BufferOwner : IDisposable
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Drawing/Utilities/ThreadLocalBlenderBuffers.cs:44]: Much neater

File: tests/ImageSharp.Drawing.Tests/Utilities/ThreadLocalBlenderBuffersTests.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+using System;
+using System.Linq;
+using System.Threading;
+using SixLabors.ImageSharp.Drawing.Utilities;
+using SixLabors.ImageSharp.PixelFormats;
+using Xunit;
+
+namespace SixLabors.ImageSharp.Drawing.Tests.Drawing.Utils
+{
+    public class ThreadLocalBlenderBuffersTests
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Drawing.Tests/Utilities/ThreadLocalBlenderBuffersTests.cs:13]: Very cool!

**Review by JimBobSquarePants — COMMENTED** [review]
This is great! Much simpler 😄

**Review by JimBobSquarePants — APPROVED** [review]

File: src/ImageSharp.Drawing/Utilities/Intersect.cs (hunk 2 above)
Comment (antonfirsov) [reply src/ImageSharp.Drawing/Utilities/Intersect.cs:14]: When necessary, I do the math on my own, in other cases I just take the steal & test approach. Here, I basically adapted the code from [here](https://stackoverflow.com/questions/563198/how-do-you-detect-where-two-line-segments-intersect/1968345#1968345), which is based on a [book](https://www.amazon.com/dp/0672323699). Anyone can do that :)

**Comment by antonfirsov** [comment]
> Update build-and-test.yml bc50f77

Thanks! 

I'll leave this open for a while to give @tocsoft a chance to react, if he has the time.

**Pull request merged**